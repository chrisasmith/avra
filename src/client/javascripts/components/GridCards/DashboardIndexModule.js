import React, { Component } from 'react'
import classnames from 'classnames'
import PropTypes from 'prop-types'
import domToImage from 'dom-to-image'
import styles from './styles/DashboardModule.scss'
import IndexGraph from './Graph/IndexGraph'
import DashboardShellModule from './DashboardShellModule'

const MAX_NUMBER_OF_VALUES = 4

const propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string,
  isFetching: PropTypes.bool.isRequired,
  data: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  scarborough: PropTypes.bool,
}

class DashboardIndexModule extends Component {
  constructor(props, context) {
    super(props, context)

    this.graph = React.createRef()

    this.onDownloadImage = this.onDownloadImage.bind(this)
    this.onDownloadCSV = this.onDownloadCSV.bind(this)
    this.onExpandClick = this.onExpandClick.bind(this)
    this.onChangeType = this.onChangeType.bind(this)

    this.state = {
      expanded: false,
      type: props.data && Array.isArray(props.data) ? props.data[0].id : null,
      isDropdownData: Array.isArray(props.data),
    }
  }

  componentWillReceiveProps(next) {
    if (next.data && !this.props.data && Array.isArray(next.data)) {
      this.setState({ type: next.data[0].id, isDropdownData: true })
    }
  }

  onDownloadImage() {
    const { title } = this.props

    domToImage.toJpeg(this.graph.current, { quality: 0.95 })
      .then((dataUrl) => {
        const link = document.createElement('a')
        link.download = `${title}.jpeg`
        link.href = dataUrl
        link.click()
      })
  }

  onDownloadCSV() {
    const { title, data } = this.props
    const { type, isDropdownData } = this.state

    const dataForType = isDropdownData ? data.find(d => d.id === type) : data

    const colATitle = ''
    const colBTitle = dataForType.name || ''

    const csvContent = dataForType.values.reduce((p, c) => {
      p += `${c.title},${c.value}`
      if (c.tooltip) {
        p += `,${c.tooltip}`
      }
      p += '\r\n'
      return p
    }, `data:text/csv;charset=utf-8,${colATitle},${colBTitle}\r\n`)

    const encodedUri = encodeURI(csvContent)
    const link = document.createElement('a')
    link.setAttribute('href', encodedUri)
    link.setAttribute('download', `${title}.csv`)
    document.body.appendChild(link)

    link.click()
  }

  onExpandClick(e) {
    if (e) e.preventDefault()

    this.setState({ expanded: !this.state.expanded })
  }

  onChangeType(e) {
    this.setState({ type: e.target.value })
  }

  render() {
    const {
      title,
      description,
      isFetching,
      data,
      scarborough,
    } = this.props

    const { isDropdownData, type } = this.state

    if (!data || isFetching) {
      return (
        <DashboardShellModule isFetching title={title} />
      )
    }

    const dataForType = isDropdownData ? data.find(d => d.id === type) : data

    const { expanded } = this.state

    // sort them
    const values = dataForType.values.sort((a, b) => b.value - a.value)

    // find the highest value
    const max = values.reduce((p, c) => p > c.value ? p : c.value, 0)

    // only show the top values
    const parsedValues = values.filter((v, i) => expanded ? true : i < MAX_NUMBER_OF_VALUES)

    // find the max and position of 1
    const chartMax = max <= 2 ? 2 : max
    const midWay = 100 / chartMax

    return (
      <DashboardShellModule
        isFetching={isFetching}
        title={title}
        onDownloadCSV={this.onDownloadCSV}
        onDownloadImage={this.onDownloadImage}
      >
        <div>
          <div className={classnames(styles.graphWrapper)}>
            <div className={classnames('p-3', styles.graphContainer)} ref={this.graph}>
              {description && (
                <div className="text-center pb-3">
                  {description}
                </div>
              )}
              {isDropdownData && (
                <div className={styles.dropdown}>
                  <select className="form-control" onChange={this.onChangeType} value={type}>
                    {data.map(t => (
                      <option key={t.id} value={t.id}>{t.name}</option>
                    ))}
                  </select>
                </div>
              )}
              <div className={styles.graphHolder}>
                <div>
                  <div>
                    {parsedValues.map((graph, idx) => (
                      <IndexGraph
                        key={idx}
                        title={graph.title}
                        tooltip={graph.tooltip}
                        value={graph.value}
                        maxValue={chartMax}
                      />
                    ))}
                  </div>
                  <div className={styles.line}>
                    <div className={styles.indicatorLine} style={{ left: `${midWay}%` }} />
                    <div className={styles.indicatorNumber} style={{ left: `${midWay}%` }}>1</div>
                  </div>
                </div>
              </div>
            </div>
            <div>
              {scarborough && (
                <div className={styles.source}>Source: Scarborough Research</div>
              )}
              {values.length > MAX_NUMBER_OF_VALUES && (
              <div className={styles.seeMoreLink}>
                <a href="#" onClick={this.onExpandClick}>{!expanded ? 'See More Results' : 'See Less Results'}</a>
              </div>
              )}
            </div>
          </div>
        </div>
      </DashboardShellModule>
    )
  }
}

DashboardIndexModule.propTypes = propTypes
export default DashboardIndexModule
