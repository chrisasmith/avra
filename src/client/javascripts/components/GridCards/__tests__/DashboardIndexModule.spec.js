import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import DashboardIndexModule from '../DashboardIndexModule'

const values = [
  { title: 'Beer', value: 1.23, tooltip: 'Have bought in the last 4 weeks' },
  { title: 'Auto', value: 1.15, tooltip: 'Intend to buy new in the next 12 months' },
  { title: 'Computer', value: 1.11, tooltip: 'Intend to buy new in the next 12 months' },
  { title: 'Computer accessories', value: 1.18, tooltip: 'Intend to buy new in the next 12 months' },
]

const arrayValues = [
  {
    id: 'all',
    name: 'All',
    values: [
      { title: 'Labatt Blue', value: 5.28 },
      { title: 'Samuel Adams', value: 2.46 },
    ],
  },
  {
    id: 'imported',
    name: 'Imported',
    values: [
      { title: 'Labatt Blue', value: 5.28 },
      { title: 'Heineken', value: 1.61 },
      { title: 'Corona Light', value: 1.52 },
    ],
  },
]

describe('DashboardIndexModule Component', () => {
  it('Should show', () => {
    const wrapper = shallow(<DashboardIndexModule
      title="title"
      isFetching={false}
      data={{ values }}
    />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('Should show scarborough', () => {
    const wrapper = shallow(<DashboardIndexModule
      title="title"
      isFetching={false}
      data={{ values }}
      scarborough
    />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('Should show with array data', () => {
    const wrapper = shallow(<DashboardIndexModule
      title="title"
      isFetching={false}
      data={arrayValues}
    />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('Should change data on dropdown', () => {
    const wrapper = shallow(<DashboardIndexModule
      title="title"
      isFetching={false}
      data={arrayValues}
    />)

    wrapper.find('select').simulate('change', { target: { value: 'imported' } })
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('Should show with a max higher than 2', () => {
    const wrapper = shallow(<DashboardIndexModule
      title="title"
      isFetching={false}
      data={{ values: [...values, { title: 'title', value: 6.18, tooltip: 'tip' }] }}
    />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('Should show expand button', () => {
    const wrapper = shallow(<DashboardIndexModule
      title="title"
      isFetching={false}
      data={{ values: [...values, { title: 'title', value: 1.18, tooltip: 'tip' }] }}
    />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('Should expand', () => {
    const wrapper = shallow(<DashboardIndexModule
      title="title"
      isFetching={false}
      data={{ values: [...values, { title: 'title', value: 1.18, tooltip: 'tip' }] }}
    />)

    wrapper.find('.seeMoreLink').find('a').simulate('click')

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('Should shrink', () => {
    const wrapper = shallow(<DashboardIndexModule
      title="title"
      isFetching={false}
      data={{ values: [...values, { title: 'title', value: 1.18, tooltip: 'tip' }] }}
    />)

    wrapper.find('.seeMoreLink').find('a').simulate('click')
    wrapper.find('.seeMoreLink').find('a').simulate('click')

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })
})
