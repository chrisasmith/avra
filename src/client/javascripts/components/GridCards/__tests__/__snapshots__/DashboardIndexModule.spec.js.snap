// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardIndexModule Component Should change data on dropdown 1`] = `
<DashboardShellModule
  isFetching={false}
  onDownloadCSV={[Function]}
  onDownloadImage={[Function]}
  title="title"
>
  <div>
    <div
      className="graphWrapper"
    >
      <div
        className="p-3 graphContainer"
      >
        <div
          className="dropdown"
        >
          <select
            className="form-control"
            onChange={[Function]}
            value="imported"
          >
            <option
              key="all"
              value="all"
            >
              All
            </option>
            <option
              key="imported"
              value="imported"
            >
              Imported
            </option>
          </select>
        </div>
        <div
          className="graphHolder"
        >
          <div>
            <div>
              <IndexGraph
                key="0"
                maxValue={5.28}
                title="Labatt Blue"
                value={5.28}
              />
              <IndexGraph
                key="1"
                maxValue={5.28}
                title="Heineken"
                value={1.61}
              />
              <IndexGraph
                key="2"
                maxValue={5.28}
                title="Corona Light"
                value={1.52}
              />
            </div>
            <div
              className="line"
            >
              <div
                className="indicatorLine"
                style={
                  Object {
                    "left": "18.939393939393938%",
                  }
                }
              />
              <div
                className="indicatorNumber"
                style={
                  Object {
                    "left": "18.939393939393938%",
                  }
                }
              >
                1
              </div>
            </div>
          </div>
        </div>
      </div>
      <div />
    </div>
  </div>
</DashboardShellModule>
`;

exports[`DashboardIndexModule Component Should expand 1`] = `
<DashboardShellModule
  isFetching={false}
  onDownloadCSV={[Function]}
  onDownloadImage={[Function]}
  title="title"
>
  <div>
    <div
      className="graphWrapper"
    >
      <div
        className="p-3 graphContainer"
      >
        <div
          className="graphHolder"
        >
          <div>
            <div>
              <IndexGraph
                key="0"
                maxValue={2}
                title="Beer"
                tooltip="Have bought in the last 4 weeks"
                value={1.23}
              />
              <IndexGraph
                key="1"
                maxValue={2}
                title="Computer accessories"
                tooltip="Intend to buy new in the next 12 months"
                value={1.18}
              />
              <IndexGraph
                key="2"
                maxValue={2}
                title="title"
                tooltip="tip"
                value={1.18}
              />
              <IndexGraph
                key="3"
                maxValue={2}
                title="Auto"
                tooltip="Intend to buy new in the next 12 months"
                value={1.15}
              />
              <IndexGraph
                key="4"
                maxValue={2}
                title="Computer"
                tooltip="Intend to buy new in the next 12 months"
                value={1.11}
              />
            </div>
            <div
              className="line"
            >
              <div
                className="indicatorLine"
                style={
                  Object {
                    "left": "50%",
                  }
                }
              />
              <div
                className="indicatorNumber"
                style={
                  Object {
                    "left": "50%",
                  }
                }
              >
                1
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          className="seeMoreLink"
        >
          <a
            href="#"
            onClick={[Function]}
          >
            See Less Results
          </a>
        </div>
      </div>
    </div>
  </div>
</DashboardShellModule>
`;

exports[`DashboardIndexModule Component Should show 1`] = `
<DashboardShellModule
  isFetching={false}
  onDownloadCSV={[Function]}
  onDownloadImage={[Function]}
  title="title"
>
  <div>
    <div
      className="graphWrapper"
    >
      <div
        className="p-3 graphContainer"
      >
        <div
          className="graphHolder"
        >
          <div>
            <div>
              <IndexGraph
                key="0"
                maxValue={2}
                title="Beer"
                tooltip="Have bought in the last 4 weeks"
                value={1.23}
              />
              <IndexGraph
                key="1"
                maxValue={2}
                title="Computer accessories"
                tooltip="Intend to buy new in the next 12 months"
                value={1.18}
              />
              <IndexGraph
                key="2"
                maxValue={2}
                title="Auto"
                tooltip="Intend to buy new in the next 12 months"
                value={1.15}
              />
              <IndexGraph
                key="3"
                maxValue={2}
                title="Computer"
                tooltip="Intend to buy new in the next 12 months"
                value={1.11}
              />
            </div>
            <div
              className="line"
            >
              <div
                className="indicatorLine"
                style={
                  Object {
                    "left": "50%",
                  }
                }
              />
              <div
                className="indicatorNumber"
                style={
                  Object {
                    "left": "50%",
                  }
                }
              >
                1
              </div>
            </div>
          </div>
        </div>
      </div>
      <div />
    </div>
  </div>
</DashboardShellModule>
`;

exports[`DashboardIndexModule Component Should show expand button 1`] = `
<DashboardShellModule
  isFetching={false}
  onDownloadCSV={[Function]}
  onDownloadImage={[Function]}
  title="title"
>
  <div>
    <div
      className="graphWrapper"
    >
      <div
        className="p-3 graphContainer"
      >
        <div
          className="graphHolder"
        >
          <div>
            <div>
              <IndexGraph
                key="0"
                maxValue={2}
                title="Beer"
                tooltip="Have bought in the last 4 weeks"
                value={1.23}
              />
              <IndexGraph
                key="1"
                maxValue={2}
                title="Computer accessories"
                tooltip="Intend to buy new in the next 12 months"
                value={1.18}
              />
              <IndexGraph
                key="2"
                maxValue={2}
                title="title"
                tooltip="tip"
                value={1.18}
              />
              <IndexGraph
                key="3"
                maxValue={2}
                title="Auto"
                tooltip="Intend to buy new in the next 12 months"
                value={1.15}
              />
            </div>
            <div
              className="line"
            >
              <div
                className="indicatorLine"
                style={
                  Object {
                    "left": "50%",
                  }
                }
              />
              <div
                className="indicatorNumber"
                style={
                  Object {
                    "left": "50%",
                  }
                }
              >
                1
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          className="seeMoreLink"
        >
          <a
            href="#"
            onClick={[Function]}
          >
            See More Results
          </a>
        </div>
      </div>
    </div>
  </div>
</DashboardShellModule>
`;

exports[`DashboardIndexModule Component Should show scarborough 1`] = `
<DashboardShellModule
  isFetching={false}
  onDownloadCSV={[Function]}
  onDownloadImage={[Function]}
  title="title"
>
  <div>
    <div
      className="graphWrapper"
    >
      <div
        className="p-3 graphContainer"
      >
        <div
          className="graphHolder"
        >
          <div>
            <div>
              <IndexGraph
                key="0"
                maxValue={2}
                title="Beer"
                tooltip="Have bought in the last 4 weeks"
                value={1.23}
              />
              <IndexGraph
                key="1"
                maxValue={2}
                title="Computer accessories"
                tooltip="Intend to buy new in the next 12 months"
                value={1.18}
              />
              <IndexGraph
                key="2"
                maxValue={2}
                title="Auto"
                tooltip="Intend to buy new in the next 12 months"
                value={1.15}
              />
              <IndexGraph
                key="3"
                maxValue={2}
                title="Computer"
                tooltip="Intend to buy new in the next 12 months"
                value={1.11}
              />
            </div>
            <div
              className="line"
            >
              <div
                className="indicatorLine"
                style={
                  Object {
                    "left": "50%",
                  }
                }
              />
              <div
                className="indicatorNumber"
                style={
                  Object {
                    "left": "50%",
                  }
                }
              >
                1
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          className="source"
        >
          Source: Scarborough Research
        </div>
      </div>
    </div>
  </div>
</DashboardShellModule>
`;

exports[`DashboardIndexModule Component Should show with a max higher than 2 1`] = `
<DashboardShellModule
  isFetching={false}
  onDownloadCSV={[Function]}
  onDownloadImage={[Function]}
  title="title"
>
  <div>
    <div
      className="graphWrapper"
    >
      <div
        className="p-3 graphContainer"
      >
        <div
          className="graphHolder"
        >
          <div>
            <div>
              <IndexGraph
                key="0"
                maxValue={6.18}
                title="title"
                tooltip="tip"
                value={6.18}
              />
              <IndexGraph
                key="1"
                maxValue={6.18}
                title="Beer"
                tooltip="Have bought in the last 4 weeks"
                value={1.23}
              />
              <IndexGraph
                key="2"
                maxValue={6.18}
                title="Computer accessories"
                tooltip="Intend to buy new in the next 12 months"
                value={1.18}
              />
              <IndexGraph
                key="3"
                maxValue={6.18}
                title="Auto"
                tooltip="Intend to buy new in the next 12 months"
                value={1.15}
              />
            </div>
            <div
              className="line"
            >
              <div
                className="indicatorLine"
                style={
                  Object {
                    "left": "16.181229773462785%",
                  }
                }
              />
              <div
                className="indicatorNumber"
                style={
                  Object {
                    "left": "16.181229773462785%",
                  }
                }
              >
                1
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          className="seeMoreLink"
        >
          <a
            href="#"
            onClick={[Function]}
          >
            See More Results
          </a>
        </div>
      </div>
    </div>
  </div>
</DashboardShellModule>
`;

exports[`DashboardIndexModule Component Should show with array data 1`] = `
<DashboardShellModule
  isFetching={false}
  onDownloadCSV={[Function]}
  onDownloadImage={[Function]}
  title="title"
>
  <div>
    <div
      className="graphWrapper"
    >
      <div
        className="p-3 graphContainer"
      >
        <div
          className="dropdown"
        >
          <select
            className="form-control"
            onChange={[Function]}
            value="all"
          >
            <option
              key="all"
              value="all"
            >
              All
            </option>
            <option
              key="imported"
              value="imported"
            >
              Imported
            </option>
          </select>
        </div>
        <div
          className="graphHolder"
        >
          <div>
            <div>
              <IndexGraph
                key="0"
                maxValue={5.28}
                title="Labatt Blue"
                value={5.28}
              />
              <IndexGraph
                key="1"
                maxValue={5.28}
                title="Samuel Adams"
                value={2.46}
              />
            </div>
            <div
              className="line"
            >
              <div
                className="indicatorLine"
                style={
                  Object {
                    "left": "18.939393939393938%",
                  }
                }
              />
              <div
                className="indicatorNumber"
                style={
                  Object {
                    "left": "18.939393939393938%",
                  }
                }
              >
                1
              </div>
            </div>
          </div>
        </div>
      </div>
      <div />
    </div>
  </div>
</DashboardShellModule>
`;

exports[`DashboardIndexModule Component Should shrink 1`] = `
<DashboardShellModule
  isFetching={false}
  onDownloadCSV={[Function]}
  onDownloadImage={[Function]}
  title="title"
>
  <div>
    <div
      className="graphWrapper"
    >
      <div
        className="p-3 graphContainer"
      >
        <div
          className="graphHolder"
        >
          <div>
            <div>
              <IndexGraph
                key="0"
                maxValue={2}
                title="Beer"
                tooltip="Have bought in the last 4 weeks"
                value={1.23}
              />
              <IndexGraph
                key="1"
                maxValue={2}
                title="Computer accessories"
                tooltip="Intend to buy new in the next 12 months"
                value={1.18}
              />
              <IndexGraph
                key="2"
                maxValue={2}
                title="title"
                tooltip="tip"
                value={1.18}
              />
              <IndexGraph
                key="3"
                maxValue={2}
                title="Auto"
                tooltip="Intend to buy new in the next 12 months"
                value={1.15}
              />
            </div>
            <div
              className="line"
            >
              <div
                className="indicatorLine"
                style={
                  Object {
                    "left": "50%",
                  }
                }
              />
              <div
                className="indicatorNumber"
                style={
                  Object {
                    "left": "50%",
                  }
                }
              >
                1
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          className="seeMoreLink"
        >
          <a
            href="#"
            onClick={[Function]}
          >
            See More Results
          </a>
        </div>
      </div>
    </div>
  </div>
</DashboardShellModule>
`;
